core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,674 @@
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
+		for (idx_jr = 0, found = false; idx_jr < jr_privdata->nb_jobs;
+		     idx_jr++) {
+			/*
+			 * Search for the caller information corresponding to
+			 * the completed JR.
+			 * Don't use the outread_index or inwrite_index because
+			 * completion can be out of order compared to input
+			 * buffer
+			 */
+			caller = &jr_privdata->callers[idx_jr];
+			if (caam_desc_pop(&jr_out->desc) == caller->pdesc) {
+				jobctx = caller->jobctx;
+				jobctx->status =
+					caam_read_jobstatus(&jr_out->status);
+
+				/* Update return Job IDs mask */
+				if (caller->job_id & wait_job_ids)
+					ret_job_id |= caller->job_id;
+
+				JR_TRACE("JR id=%d, context @0x%08" PRIxVA,
+					 caller->job_id, (vaddr_t)jobctx);
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc  = 0;
+				caller->job_id = JR_JOB_FREE;
+				found = true;
+				JR_TRACE("Free space #%d in the callers array",
+					 idx_jr);
+				break;
+			}
+		}
+		cpu_spin_unlock(&jr_privdata->callers_lock);
+
+		/*
+		 * Remove the JR from the output list even if no
+		 * JR caller found
+		 */
+		caam_hal_jr_del_job(jr_privdata->baseaddr);
+
+		/*
+		 * Increment index to next JR output entry taking care that
+		 * is a circular buffer of nb_jobs size.
+		 */
+		jr_privdata->outread_index++;
+		jr_privdata->outread_index %= jr_privdata->nb_jobs;
+
+		if (found && jobctx->callback) {
+			/* Finally, execute user's callback */
+			jobctx->callback(jobctx);
+		}
+
+	} while (--nb_jobs_done);
+
+	cpu_spin_unlock_xrestore(&jr_privdata->outlock, exceptions);
+
+	return ret_job_id;
+}
+
+/*
+ * Enqueues a new job in the Job Ring input queue. Keep the caller's
+ * job context in private array.
+ *
+ * @jobctx   Caller's job context
+ * @job_id   [out] Job ID enqueued
+ */
+static enum caam_status do_jr_enqueue(struct caam_jobctx *jobctx,
+				      uint32_t *job_id)
+{
+	enum caam_status retstatus = CAAM_BUSY;
+	struct inring_entry *cur_inrings = NULL;
+	struct caller_info *caller = NULL;
+	uint32_t exceptions = 0;
+	uint32_t job_mask = 0;
+	uint8_t idx_jr = 0;
+	bool found = false;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->inlock);
+
+	/*
+	 * Stay lock until a job is available
+	 * Check if there is an available JR index in the HW
+	 */
+	while (caam_hal_jr_read_nbslot_available(jr_privdata->baseaddr) == 0) {
+		/*
+		 * WFE will return thanks to a SEV generated by the
+		 * interrupt handler or by a spin_unlock
+		 */
+		wfe();
+	};
+
+	/*
+	 * There is a space free in the input ring but it doesn't mean
+	 * that the job pushed is completed.
+	 * Completion is out of order. Look for a free space in the
+	 * caller data to push them and get a job ID for the completion
+	 *
+	 * Lock the caller information array because dequeue is
+	 * also touching it
+	 */
+	cpu_spin_lock(&jr_privdata->callers_lock);
+	for (idx_jr = 0; idx_jr < jr_privdata->nb_jobs && !found; idx_jr++) {
```

Comment:
nitpicking: here extracting `!found` from the loop control would be nice. Prefer a `break` at line 363.

Reply:
ok